### ISSUE TITLE
cromwell slows down in large scatters (against sge)

### ISSUE THREAD

**Issue by yfarjoun** [broadinstitute/cromwell#1938]
I have a 1300 shard scatter and while cromwell started off quickly, it is now submitting jobs at a rate of 1/minute (!!!!!) the jstack is showing that only one thread is running with the following trace (see below).

This doesn't seem to happen against GCS.

Also, of note: I'm running without a database.

the wdl (selfcontained) is here:

```
task Decapitate {
	File inputFile
	String outputFilename

	command {
		tail -n+2 ${inputFile}
	}
	runtime {
  		memory: "4 GB"
  	}
	output  {
		Array[Array[String]] outputMatrix=read_tsv(stdout())
	}
}

task BreakUpRow {
	String in_SAMPLE_ALIAS
	String in_PROJECT
	String in_DATA_TYPE
	String in_COMPARE_SAMPLE
	String in_COMPARE_PROJECT
	String in_CLEAN_SAMPLE
	String in_CLEAN_COMPARE_SAMPLE
	String in_FILE
	String in_CONTAM

	command {
		#do nothing
	}
	runtime {
  		memory: "1 GB"
  	}
	output {
		String SAMPLE_ALIAS = "${in_SAMPLE_ALIAS}"
		String PROJECT = "${in_PROJECT}"
		String DATA_TYPE = "${in_DATA_TYPE}"
		String COMPARE_SAMPLE = "${in_COMPARE_SAMPLE}"
		String COMPARE_PROJECT = "${in_COMPARE_PROJECT}"
		String CLEAN_SAMPLE = "${in_CLEAN_SAMPLE}"
		String CLEAN_COMPARE_SAMPLE = "${in_CLEAN_COMPARE_SAMPLE}"
		String FILE = "${in_FILE}"
		String CONTAM = "${in_CONTAM}"
	}

}

task ExtractContaminant {
	String picard_jar
	File input_bam
	String contaminant_sample
	File haplotype_database
	File ref_fasta
	File ref_fasta_idx
	File ref_dict
	String contamination_rate

	command {
		java -jar ${picard_jar} IdentifyContaminant \
			I=${input_bam} \
			O=${contaminant_sample}.vcf \
			H=${haplotype_database} \
			R=${ref_fasta} \
			C=${contamination_rate} \
			SAMPLE_ALIAS=${contaminant_sample}
	}
	runtime {
  		memory: "4 GB"
  	}	
	output {
		File vcf = "${contaminant_sample}.vcf"
		File vcf_idx = "${contaminant_sample}.vcf.idx"
	}
}

task Fingerprint {
  String PICARD
  File input_vcf
  File input_vcf_index
  File haplotype_database_file
  File genotypes
  String sample_name
  String output_name
 
  command {
    java -Dsamjdk.buffer_size=131072 -XX:GCTimeLimit=50 -XX:GCHeapFreeLimit=10 -Xmx1024m  \
    -jar ${PICARD} \
    CheckFingerprint \
    INPUT=${input_vcf} \
    OUTPUT=${output_name} \
    GENOTYPES=${genotypes} \
    HAPLOTYPE_MAP=${haplotype_database_file} \
    SAMPLE_ALIAS="${sample_name}"
  }
  runtime {
  	continueOnReturnCode: true
  	memory: "4 GB"
  }
  output {
    File summary_metrics = "${output_name}.fingerprinting_summary_metrics"
    File detail_metrics = "${output_name}.fingerprinting_detail_metrics"
  }
}

workflow fingerprintContaminants {
	File SamplesTSV = "/dsde/working/farjoun/AGBT2017/SampleSetWithcontamination.txt"
	String picard_jar = "/seq/software/picard/current/bin/picard-private.jar"
	File haplotype_database = "/seq/references/Homo_sapiens_assembly19/v1/Homo_sapiens_assembly19.haplotype_database.txt"
	File ref_fasta = "/seq/references/Homo_sapiens_assembly19/v1/Homo_sapiens_assembly19.fasta"
	File ref_fasta_idx = "/seq/references/Homo_sapiens_assembly19/v1/Homo_sapiens_assembly19.fasta.fai"
	File ref_dict = "/seq/references/Homo_sapiens_assembly19/v1/Homo_sapiens_assembly19.dict"

	call Decapitate {
		input:
			inputFile=SamplesTSV,
			outputFilename=SamplesTSV+".headless"
	}

	scatter(row in Decapitate.outputMatrix){
		call BreakUpRow {
			input:
				in_SAMPLE_ALIAS = row[0],
				in_PROJECT = row[1],
				in_DATA_TYPE = row[2],
				in_COMPARE_SAMPLE = row[3],
				in_COMPARE_PROJECT = row[4],
				in_CLEAN_SAMPLE = row[5],
				in_CLEAN_COMPARE_SAMPLE = row[6],
				in_FILE = row[7],
				in_CONTAM = row[8]
		}
		
		call ExtractContaminant {
			input:
				picard_jar = picard_jar,
				input_bam = BreakUpRow.FILE,
				contaminant_sample = BreakUpRow.CLEAN_COMPARE_SAMPLE,
				contamination_rate = BreakUpRow.CONTAM,
				haplotype_database = haplotype_database,
				ref_fasta = ref_fasta,
				ref_fasta_idx = ref_fasta_idx,
				ref_dict = ref_dict
		}

		call Fingerprint {
			input:
				PICARD = picard_jar,
  				input_vcf = ExtractContaminant.vcf,
  				input_vcf_index = ExtractContaminant.vcf_idx,
  				haplotype_database_file = haplotype_database,
  				genotypes = "/seq/references/reference_genotypes/non-hapmap/" + BreakUpRow.COMPARE_PROJECT + "/Homo_sapiens_assembly19/" + BreakUpRow.CLEAN_COMPARE_SAMPLE + ".vcf",
  				sample_name = BreakUpRow.CLEAN_COMPARE_SAMPLE,
  				output_name = BreakUpRow.CLEAN_COMPARE_SAMPLE + "." + BreakUpRow.CONTAM + "." + BreakUpRow.DATA_TYPE 
		}
	}
}
```

the stack trace is here

```



  java.lang.Thread.State: RUNNABLE
	at scala.collection.mutable.AbstractBuffer.<init>(Buffer.scala:49)
	at scala.collection.mutable.ListBuffer.<init>(ListBuffer.scala:46)
	at scala.collection.immutable.List$.newBuilder(List.scala:453)
	at scala.collection.generic.GenericTraversableTemplate$class.genericBuilder(GenericTraversableTemplate.scala:70)
	at scala.collection.AbstractTraversable.genericBuilder(Traversable.scala:104)
	at scala.collection.generic.GenTraversableFactory$GenericCanBuildFrom.apply(GenTraversableFactory.scala:57)
	at scala.collection.generic.GenTraversableFactory$GenericCanBuildFrom.apply(GenTraversableFactory.scala:52)
	at scala.collection.SeqLike$class.$colon$plus(SeqLike.scala:556)
	at scala.collection.AbstractSeq.$colon$plus(Seq.scala:41)
	at wdl4s.Scope$class.fullyQualifiedName(Scope.scala:92)
	at wdl4s.Call.fullyQualifiedName(Call.scala:60)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$cromwell$engine$workflow$lifecycle$execution$ExecutionStore$$isDone$1$1.apply(ExecutionStore.scala:94)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$cromwell$engine$workflow$lifecycle$execution$ExecutionStore$$isDone$1$1.apply(ExecutionStore.scala:93)
	at scala.collection.Iterator$class.exists(Iterator.scala:919)
	at scala.collection.AbstractIterator.exists(Iterator.scala:1336)
	at scala.collection.IterableLike$class.exists(IterableLike.scala:77)
	at scala.collection.AbstractIterable.exists(Iterable.scala:54)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore.cromwell$engine$workflow$lifecycle$execution$ExecutionStore$$isDone$1(ExecutionStore.scala:93)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$5.apply(ExecutionStore.scala:98)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$5.apply(ExecutionStore.scala:98)
	at scala.collection.Iterator$class.forall(Iterator.scala:905)
	at scala.collection.AbstractIterator.forall(Iterator.scala:1336)
	at scala.collection.IterableLike$class.forall(IterableLike.scala:75)
	at scala.collection.AbstractIterable.forall(Iterable.scala:54)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore.arePrerequisitesDone(ExecutionStore.scala:98)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore.cromwell$engine$workflow$lifecycle$execution$ExecutionStore$$isRunnable(ExecutionStore.scala:43)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$runnableScopes$1.applyOrElse(ExecutionStore.scala:37)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$runnableScopes$1.applyOrElse(ExecutionStore.scala:37)
	at scala.collection.immutable.List.collect(List.scala:303)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore.runnableScopes(ExecutionStore.scala:37)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActor$$startRunnableScopes(WorkflowExecutionActor.scala:347)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.handleExecutionSuccess(WorkflowExecutionActor.scala:326)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActor$$handleCallSuccessful(WorkflowExecutionActor.scala:304)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor$$anonfun$3.applyOrElse(WorkflowExecutionActor.scala:97)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor$$anonfun$3.applyOrElse(WorkflowExecutionActor.scala:82)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
	at akka.actor.FSM$class.processEvent(FSM.scala:663)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.akka$actor$LoggingFSM$$super$processEvent(WorkflowExecutionActor.scala:33)
	at akka.actor.LoggingFSM$class.processEvent(FSM.scala:799)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.processEvent(WorkflowExecutionActor.scala:33)
	at akka.actor.FSM$class.akka$actor$FSM$$processMsg(FSM.scala:657)
	at akka.actor.FSM$$anonfun$receive$1.applyOrElse(FSM.scala:651)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:484)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.aroundReceive(WorkflowExecutionActor.scala:33)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)


```

**Comment by geoffjentry** [comment]
@yfarjoun You're always using a DB, so what you're doing is using an in memory DB. Note that I've found that MySQL gives better performance. When you run against GCS are you also using the default in memory DB?

Are they both the same version? If so, what version. If develop, from when?

I did recently make some changes to develop that should tremendously help what this *might* be but there are also spots which are a lot more single threaded than one would like and it's possible you've found yourself in a situation where that's true on the SFS backend but not JES.

**Comment by yfarjoun** [comment]
version 24.
didn't run this actual wdl in cloud due to the dependence on many (MANY!) local files.

another thing I found was that there might be little "object storage" space left: 
```
Heap
 PSYoungGen      total 1966592K, used 891894K [0x0000000580100000, 0x00000006a7d80000, 0x0000000800000000)
  eden space 1965568K, 45% used [0x0000000580100000,0x00000005b6715b20,0x00000005f8080000)
  from space 1024K, 90% used [0x00000006a7c80000,0x00000006a7d68000,0x00000006a7d80000)
  to   space 1536K, 0% used [0x00000006a7a80000,0x00000006a7a80000,0x00000006a7c00000)
 ParOldGen       total 2513408K, used 2436406K [0x0000000080200000, 0x0000000119880000, 0x0000000580100000)
  object space 2513408K, 96% used [0x0000000080200000,0x0000000114d4db58,0x0000000119880000)
 Metaspace       used 54568K, capacity 55776K, committed 56112K, reserved 1097728K
  class space    used 7106K, capacity 7334K, committed 7472K, reserved 1048576K
```

**Comment by yfarjoun** [comment]
perhaps I should run the java with a larger -Xmx?

**Comment by meganshand** [comment]
FWIW, I've had similar symptoms while using a MySQL DB.

**Comment by geoffjentry** [comment]
It's worth a try. I've found that big (for a relatively modest "big") scatters of file paths end up chewing a lot of memory. In particular if you're globbing at all there are definitely some single threaded choke points (which is fine in a system running a ton of WFs, not so good if you're the only one) around the processing of those files.

For both you and @meganshand it's worth trying upgrading to develop and making sure to include `rewriteBatchedStatements=true` to your JDBC URL. I think this only helps on MySQL however, I remember having reason to believe it didn't do anything on HSQL but I didn't try it.

**Comment by meganshand** [comment]
I'm trying it with 16g right now. It still slows down abruptly after completing the first task inside of the scatter. The first task is fast, but then the second task (which depends on the first) is slower.

**Comment by meganshand** [comment]
Sorry, to be clear, the second task itself is not slow, but submitting it is.

**Comment by geoffjentry** [comment]
@yfarjoun how stable is that stacktrace? IOW if you do a thread dump repeatedly is it generally hanging in that area?

**Comment by meganshand** [comment]
I tried to throttle to only 50 jobs at a time, but it didn't seem to help.

**Comment by cjllanwarne** [comment]
To confirm, this is definitely not just the "call caching via md5 is slow" scenario?

**Comment by yfarjoun** [comment]
@geoffjentry pretty stable. i did ctrl-\ about 5 times and it always landed in the same place.
@cjllanwarne @meganshand is using md5 on paths, not content iianm

**Comment by kcibul** [comment]
@yfarjoun are you saying way back that this is specific to the SFS backend? (ie not GCS)?

Adding to the Joint Calling milestone... we've got to be able to scale at least 3k wide

**Comment by geoffjentry** [comment]
@kcibul based on the stack trace I suspect this is generic, despite the report of sge-specific. There was a suspiciously single threaded operation going on there IIRC but if could be one of the many cases where it's a good thing in a multi-workflow system but terrible in a single one

**Comment by geoffjentry** [comment]
Yeah, a github-based scan of the code just now does make me think that this is a workflow-level slowdown, which again is total crap if you're the only workflow in the system. I don't know why only 3k wide would manifest here though, I've run things 200k wide and not seen a problem in this spot, and I don't think the processing here would be a function of the inputs/outputs/etc 

What I was getting at with the single vs multi-workflow thing though is that many of the cases like this that I've seen the solution is to parallelize the operation, but in a system running thousands of workflows you don't just want to spam a ton of threads to do this as you'll choke out everyone else. In contrast if you're a single workflow on an n-core box you'd want to use as many cores as you can for this. In the general case I don't know how to resolve that other than to set up a special execution context for these sorts of things, default them to having a a bunch of CPU and then letting someone like firecloud configure that down.

That said, in this particular area we've previously seen some highly inefficient copying/comparisons going on so that's also a possible culprit here.

**Comment by yfarjoun** [comment]
hmmm. I'm not call-caching (enabled=false) and we are using links. 

As an interesting observation, the workflow managed to get through the second step in the scatter, and when it started the third step it sped up, submitting jobs at about 7sec / job.

**Comment by yfarjoun** [comment]
the stack, however, remianed the same (?) 

```
java.lang.Thread.State: RUNNABLE
	at java.util.Arrays.copyOfRange(Arrays.java:3658)
	at java.lang.String.<init>(String.java:201)
	at java.lang.StringBuilder.toString(StringBuilder.java:407)
	at scala.collection.mutable.StringBuilder.toString(StringBuilder.scala:430)
	at scala.collection.TraversableOnce$class.mkString(TraversableOnce.scala:323)
	at scala.collection.AbstractTraversable.mkString(Traversable.scala:104)
	at scala.collection.TraversableOnce$class.mkString(TraversableOnce.scala:325)
	at scala.collection.AbstractTraversable.mkString(Traversable.scala:104)
	at wdl4s.Scope$class.fullyQualifiedName(Scope.scala:92)
	at wdl4s.Declaration.fullyQualifiedName(Declaration.scala:97)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$cromwell$engine$workflow$lifecycle$execution$ExecutionStore$$isDone$1$1.apply(ExecutionStore.scala:94)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$cromwell$engine$workflow$lifecycle$execution$ExecutionStore$$isDone$1$1.apply(ExecutionStore.scala:93)
	at scala.collection.Iterator$class.exists(Iterator.scala:919)
	at scala.collection.AbstractIterator.exists(Iterator.scala:1336)
	at scala.collection.IterableLike$class.exists(IterableLike.scala:77)
	at scala.collection.AbstractIterable.exists(Iterable.scala:54)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore.cromwell$engine$workflow$lifecycle$execution$ExecutionStore$$isDone$1(ExecutionStore.scala:93)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$5.apply(ExecutionStore.scala:98)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$5.apply(ExecutionStore.scala:98)
	at scala.collection.Iterator$class.forall(Iterator.scala:905)
	at scala.collection.AbstractIterator.forall(Iterator.scala:1336)
	at scala.collection.IterableLike$class.forall(IterableLike.scala:75)
	at scala.collection.AbstractIterable.forall(Iterable.scala:54)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore.arePrerequisitesDone(ExecutionStore.scala:98)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore.cromwell$engine$workflow$lifecycle$execution$ExecutionStore$$isRunnable(ExecutionStore.scala:43)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$runnableScopes$1.applyOrElse(ExecutionStore.scala:37)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$runnableScopes$1.applyOrElse(ExecutionStore.scala:37)
	at scala.collection.immutable.List.collect(List.scala:303)
	at cromwell.engine.workflow.lifecycle.execution.ExecutionStore.runnableScopes(ExecutionStore.scala:37)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActor$$startRunnableScopes(WorkflowExecutionActor.scala:347)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.handleExecutionSuccess(WorkflowExecutionActor.scala:326)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActor$$handleCallSuccessful(WorkflowExecutionActor.scala:304)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor$$anonfun$3.applyOrElse(WorkflowExecutionActor.scala:97)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor$$anonfun$3.applyOrElse(WorkflowExecutionActor.scala:82)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
	at akka.actor.FSM$class.processEvent(FSM.scala:663)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.akka$actor$LoggingFSM$$super$processEvent(WorkflowExecutionActor.scala:33)
	at akka.actor.LoggingFSM$class.processEvent(FSM.scala:799)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.processEvent(WorkflowExecutionActor.scala:33)
	at akka.actor.FSM$class.akka$actor$FSM$$processMsg(FSM.scala:657)
	at akka.actor.FSM$$anonfun$receive$1.applyOrElse(FSM.scala:651)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:484)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.aroundReceive(WorkflowExecutionActor.scala:33)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by yfarjoun** [comment]
another point: ls takes about 15 seconds to complete where I'm running this (/dsde/working/farjoun/AGBT2017) could this the problem?

**Comment by yfarjoun** [comment]
I was told to try the cromwell_24_hotfix, but I can't build scala and I don't have the time to set this up. can someone point me to a jar I can use?

**Comment by yfarjoun** [comment]
is this related to long overheads I'm seeing in a 900 way scatter in firecloud (as I see that firecloud is on 24 now...)

https://portal.firecloud.org/#workspaces/broad-ccdg-dev%3AFunctionallyEquivalent-CCDG/Monitor/97002137-7006-47a6-90c8-faf471f0b2d1

**Comment by yfarjoun** [comment]
<img width="1294" alt="screen shot 2017-02-02 at 5 35 49 pm" src="https://cloud.githubusercontent.com/assets/2745554/22571529/16007020-e96e-11e6-9f07-4240fe4166a4.png">

**Comment by yfarjoun** [comment]
you can't see it, but the scale is 5 hours side to side...

**Comment by geoffjentry** [comment]
Another note to add that while i can't reproduce Yossi's error, he & I had previously identified `Scope.fullyQualifiedName` as a possible culprit. I decided to look at why MWDA is so much slower and that *is* being gated by the same function, to the tune of (at the time of this writing) 84.6% (and rising) of the total runtime so far.

**Comment by geoffjentry** [comment]
Hah, mea culpa! That *is* Yossi's error!

I misread the logs earlier and had been unwittingly sitting at the same step that he's on. It just takes a lot wider of a scatter on my machine than the one he's on apparently.

We have reason to believe that that value can't be fully static, but it might be able to be locked in after construction.

**Comment by yfarjoun** [comment]
this sounds promising. Can you comment on the overhead I'm seeing in firecloud? is it related?

**Comment by geoffjentry** [comment]
no idea

**Comment by yfarjoun** [comment]
I've run using the snapshot @geoffjentry gave me and while I don't think I see improvement (still 10sec per submittion to sge for the second part of the scatter) the stack trace looks different, I think:

```
   java.lang.Thread.State: RUNNABLE
	at scala.collection.Iterator$class.exists(Iterator.scala:919)
	at scala.collection.AbstractIterator.exists(Iterator.scala:1336)
	at scala.collection.IterableLike$class.exists(IterableLike.scala:77)
	at scala.collection.AbstractIterable.exists(Iterable.scala:54)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData$$anonfun$cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActorData$$upstreamFailed$1$1.apply(WorkflowExecutionActorData.scala:88)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData$$anonfun$cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActorData$$upstreamFailed$1$1.apply(WorkflowExecutionActorData.scala:87)
	at scala.collection.TraversableLike$$anonfun$filterImpl$1.apply(TraversableLike.scala:248)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.TraversableLike$class.filterImpl(TraversableLike.scala:247)
	at scala.collection.TraversableLike$class.filter(TraversableLike.scala:259)
	at scala.collection.AbstractTraversable.filter(Traversable.scala:104)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData.cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActorData$$upstreamFailed$1(WorkflowExecutionActorData.scala:87)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData$$anonfun$3.apply(WorkflowExecutionActorData.scala:93)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData$$anonfun$3.apply(WorkflowExecutionActorData.scala:92)
	at scala.collection.TraversableLike$$anonfun$filterImpl$1.apply(TraversableLike.scala:248)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.TraversableLike$class.filterImpl(TraversableLike.scala:247)
	at scala.collection.TraversableLike$class.filter(TraversableLike.scala:259)
	at scala.collection.AbstractTraversable.filter(Traversable.scala:104)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData.workflowCompletionStatus(WorkflowExecutionActorData.scala:92)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.handleExecutionSuccess(WorkflowExecutionActor.scala:323)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActor$$handleCallSuccessful(WorkflowExecutionActor.scala:308)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor$$anonfun$3.applyOrElse(WorkflowExecutionActor.scala:96)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor$$anonfun$3.applyOrElse(WorkflowExecutionActor.scala:81)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
	at akka.actor.FSM$class.processEvent(FSM.scala:663)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.akka$actor$LoggingFSM$$super$processEvent(WorkflowExecutionActor.scala:32)
	at akka.actor.LoggingFSM$class.processEvent(FSM.scala:799)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.processEvent(WorkflowExecutionActor.scala:32)
	at akka.actor.FSM$class.akka$actor$FSM$$processMsg(FSM.scala:657)
	at akka.actor.FSM$$anonfun$receive$1.applyOrElse(FSM.scala:651)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:496)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.aroundReceive(WorkflowExecutionActor.scala:32)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by yfarjoun** [comment]
I just checked again. This stacktrace seems stable for the snapshot (cromwell-25-1d61047-SNAP)

**Comment by geoffjentry** [comment]
Cool thanks. Like I said it's be a couple of days before I can look at what, if anything that affected. When I do I'll look at the profiler again but it's pretty likely that we found a, not the, bottleneck

**Comment by yfarjoun** [comment]
well, I'm happy to keep re-running my job if that helps you smooth out the kinks.

**Comment by yfarjoun** [comment]
(in the meantime cromwell in firecloud seems to have huge overhead...I wonder if it's related.)

**Comment by geoffjentry** [comment]
I'll want to use the profiler before tweaking anything else but I'll give you something when I can :)

**Comment by geoffjentry** [comment]
Don't forget that there are all sorts of extra limitations there, e.g. There is a maximum rate at which one can poll jobs due to qps limitations.

**Comment by yfarjoun** [comment]
yeah...I don't think that 5 hours overhead is reasonable even with "extra limitations"....

**Comment by geoffjentry** [comment]
I'm not trying to say that's it (I haven't thought at all about the FC case) but it's possible. What I saw was that'd it would often take a half hour to an hour to clear out a 40-50k pool of jobs due to quotas. Now imagine if there were hundreds of thousands of jobs being tracked by FC, the problem is way worse.

2 things: I doubt that this was the cas in your example and yes google is aware that this sucks.

**Comment by geoffjentry** [comment]
Got a chance to look w/ a profiler. Situation definitely *is* improved over previous in that I had to dramatically increase the size of the scatter to get hung up again (this is on my laptop which is apparently way faster than whatever machine @yfarjoun is using).

The new bottleneck appears to be ExecutionStore.arePrerequisitesDone, sitting at 99% and growing CPU usage. Specifically the `exists` call in ExecutionStore.isDone and the collect on `key.scope.upstream`. 

Note that `isDone` was also the previous hotspot but it doesn't appear to be the FQN calculation any more, rather just the `exists` itself. 

It's possible that there's still something we can do a la the FQN here but if not my concern is that this is going to take you into the "something clever" realm.

BTW @yfarjoun whatever machine you're running this on is also part of your bottleneck. I was able to do a 40k scatter no problem on one of my laptop cores, then just threw in 200k which is what locked it up. If you can't do 1k perhaps retry on something not from the stone age? ;)

**Comment by geoffjentry** [comment]
A convo with @mcovarr led me to this issue which appears to be related: https://github.com/broadinstitute/cromwell/issues/1602

It links another issue (attn @yfarjoun - look familiar?): https://github.com/broadinstitute/cromwell/issues/1603

**Comment by yfarjoun** [comment]
I was running this on gsa4. Can you submit sge jobs from your laptop? can
your sctter test on gsa4 submitting to sge before you insult my beautiful
iMac? :-p

**Comment by geoffjentry** [comment]
Maybe we can paint some racing stripes on the side of gsa4

**Comment by yfarjoun** [comment]
sure. I'll suggest that to BITS next time I see them...

**Comment by vdauwera** [comment]
Uh, isn't gsa4 supposed to be reserved for GATK automated test suites and release machinery? Please don't use it as an experimental pod racer or anything like that. If you take it down it affects user-facing systems.

**Comment by yfarjoun** [comment]
lets discuss this in a different channel.

**Comment by pgrosu** [comment]
You might be fighting your garbage collector, which might causing all of this on one machine.

**Comment by geoffjentry** [comment]
Hi @pgrosu I don't think that's the case here (didn't appear to be any warning signs on the telemetry) but there are cases where I am suspicious of that, so it's good to keep that in mind. Thanks!

**Comment by katevoss** [comment]
I've lowered the priority on this bug as Yossi has a workaround and we are working on it in parallel.

**Comment by meganshand** [comment]
I'm trying to run the same wdl on gsa5 now with the snapshot. I tried it with call caching and it is slow to retrieve the cache hits. One every 5 seconds or so. Here's the relevant jstack:

```
"cromwell-system-akka.dispatchers.engine-dispatcher-34" #99 prio=5 os_prio=0 tid=0x00002b4fa8024800 nid=0x174b2 runnable [0x00002b4da8502000]
   java.lang.Thread.State: RUNNABLE
	at scala.collection.Iterator$class.exists(Iterator.scala:919)
	at scala.collection.AbstractIterator.exists(Iterator.scala:1336)
	at scala.collection.IterableLike$class.exists(IterableLike.scala:77)
	at scala.collection.AbstractIterable.exists(Iterable.scala:54)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData$$anonfun$cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActorData$$upstreamFailed$1$1.apply(WorkflowExecutionActorData.scala:88)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData$$anonfun$cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActorData$$upstreamFailed$1$1.apply(WorkflowExecutionActorData.scala:87)
	at scala.collection.TraversableLike$$anonfun$filterImpl$1.apply(TraversableLike.scala:248)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.TraversableLike$class.filterImpl(TraversableLike.scala:247)
	at scala.collection.TraversableLike$class.filter(TraversableLike.scala:259)
	at scala.collection.AbstractTraversable.filter(Traversable.scala:104)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData.cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActorData$$upstreamFailed$1(WorkflowExecutionActorData.scala:87)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData$$anonfun$3.apply(WorkflowExecutionActorData.scala:93)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData$$anonfun$3.apply(WorkflowExecutionActorData.scala:92)
	at scala.collection.TraversableLike$$anonfun$filterImpl$1.apply(TraversableLike.scala:248)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.TraversableLike$class.filterImpl(TraversableLike.scala:247)
	at scala.collection.TraversableLike$class.filter(TraversableLike.scala:259)
	at scala.collection.AbstractTraversable.filter(Traversable.scala:104)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData.workflowCompletionStatus(WorkflowExecutionActorData.scala:92)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.handleExecutionSuccess(WorkflowExecutionActor.scala:323)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActor$$handleCallSuccessful(WorkflowExecutionActor.scala:308)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor$$anonfun$3.applyOrElse(WorkflowExecutionActor.scala:96)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor$$anonfun$3.applyOrElse(WorkflowExecutionActor.scala:81)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
	at akka.actor.FSM$class.processEvent(FSM.scala:663)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.akka$actor$LoggingFSM$$super$processEvent(WorkflowExecutionActor.scala:32)
	at akka.actor.LoggingFSM$class.processEvent(FSM.scala:799)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.processEvent(WorkflowExecutionActor.scala:32)
	at akka.actor.FSM$class.akka$actor$FSM$$processMsg(FSM.scala:657)
	at akka.actor.FSM$$anonfun$receive$1.applyOrElse(FSM.scala:651)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:496)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.aroundReceive(WorkflowExecutionActor.scala:32)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by meganshand** [comment]
I tried the same thing without call caching on gsa5, same version of cromwell as above. It's much faster, but still gets slower throughout the workflow. The scatter is made of three jobs each depends on the previous one. The first job, is submitted very quickly to SGE (same job was taking 5 secs per job to retrieve from call caching as per previous note). The second job is submitting faster than before, at 5 secs per job, with the following jstack:

```
"cromwell-system-akka.dispatchers.engine-dispatcher-100" #166 prio=5 os_prio=0 tid=0x00002b59440ad000 nid=0x17ac7 runnable [0x00002b5b700fe000]
   java.lang.Thread.State: RUNNABLE
	at scala.collection.Iterator$class.exists(Iterator.scala:919)
	at scala.collection.AbstractIterator.exists(Iterator.scala:1336)
	at scala.collection.IterableLike$class.exists(IterableLike.scala:77)
	at scala.collection.AbstractIterable.exists(Iterable.scala:54)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData$$anonfun$cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActorData$$upstreamFailed$1$1.apply(WorkflowExecutionActorData.scala:88)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData$$anonfun$cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActorData$$upstreamFailed$1$1.apply(WorkflowExecutionActorData.scala:87)
	at scala.collection.TraversableLike$$anonfun$filterImpl$1.apply(TraversableLike.scala:248)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.TraversableLike$class.filterImpl(TraversableLike.scala:247)
	at scala.collection.TraversableLike$class.filter(TraversableLike.scala:259)
	at scala.collection.AbstractTraversable.filter(Traversable.scala:104)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData.cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActorData$$upstreamFailed$1(WorkflowExecutionActorData.scala:87)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData$$anonfun$3.apply(WorkflowExecutionActorData.scala:93)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData$$anonfun$3.apply(WorkflowExecutionActorData.scala:92)
	at scala.collection.TraversableLike$$anonfun$filterImpl$1.apply(TraversableLike.scala:248)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.TraversableLike$class.filterImpl(TraversableLike.scala:247)
	at scala.collection.TraversableLike$class.filter(TraversableLike.scala:259)
	at scala.collection.AbstractTraversable.filter(Traversable.scala:104)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActorData.workflowCompletionStatus(WorkflowExecutionActorData.scala:92)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.handleExecutionSuccess(WorkflowExecutionActor.scala:323)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActor$$handleCallSuccessful(WorkflowExecutionActor.scala:308)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor$$anonfun$3.applyOrElse(WorkflowExecutionActor.scala:96)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor$$anonfun$3.applyOrElse(WorkflowExecutionActor.scala:81)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
	at akka.actor.FSM$class.processEvent(FSM.scala:663)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.akka$actor$LoggingFSM$$super$processEvent(WorkflowExecutionActor.scala:32)
	at akka.actor.LoggingFSM$class.processEvent(FSM.scala:799)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.processEvent(WorkflowExecutionActor.scala:32)
	at akka.actor.FSM$class.akka$actor$FSM$$processMsg(FSM.scala:657)
	at akka.actor.FSM$$anonfun$receive$1.applyOrElse(FSM.scala:651)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:496)
	at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.aroundReceive(WorkflowExecutionActor.scala:32)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by geoffjentry** [comment]
There was a firecloud workflow with the following this morning. I suspect it's related:

```"cromwell-system-akka.dispatchers.engine-dispatcher-521" #33028 prio=5 os_prio=0 tid=0x00007efff40fd800 nid=0x386 runnable [0x00007effde1e0000]
   java.lang.Thread.State: RUNNABLE
    at scala.collection.immutable.TrieIterator.next0(TrieIterator.scala:135)
    at scala.collection.immutable.TrieIterator.next(TrieIterator.scala:103)
    at scala.collection.Iterator$class.exists(Iterator.scala:919)
    at scala.collection.AbstractIterator.exists(Iterator.scala:1336)
    at scala.collection.IterableLike$class.exists(IterableLike.scala:77)
    at scala.collection.AbstractIterable.exists(Iterable.scala:54)
    at cromwell.engine.workflow.lifecycle.execution.ExecutionStore.cromwell$engine$workflow$lifecycle$execution$ExecutionStore$$isDone$1(ExecutionStore.scala:93)
    at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$5.apply(ExecutionStore.scala:98)
    at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$5.apply(ExecutionStore.scala:98)
    at scala.collection.Iterator$class.forall(Iterator.scala:905)
    at scala.collection.AbstractIterator.forall(Iterator.scala:1336)
    at scala.collection.IterableLike$class.forall(IterableLike.scala:75)
    at scala.collection.AbstractIterable.forall(Iterable.scala:54)
    at cromwell.engine.workflow.lifecycle.execution.ExecutionStore.arePrerequisitesDone(ExecutionStore.scala:98)
    at cromwell.engine.workflow.lifecycle.execution.ExecutionStore.cromwell$engine$workflow$lifecycle$execution$ExecutionStore$$isRunnable(ExecutionStore.scala:43)
    at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$runnableScopes$1.applyOrElse(ExecutionStore.scala:37)
    at cromwell.engine.workflow.lifecycle.execution.ExecutionStore$$anonfun$runnableScopes$1.applyOrElse(ExecutionStore.scala:37)
    at scala.collection.immutable.List.collect(List.scala:295)
    at cromwell.engine.workflow.lifecycle.execution.ExecutionStore.runnableScopes(ExecutionStore.scala:37)
    at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActor$$startRunnableScopes(WorkflowExecutionActor.scala:357)
    at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.handleExecutionSuccess(WorkflowExecutionActor.scala:336)
    at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.cromwell$engine$workflow$lifecycle$execution$WorkflowExecutionActor$$handleCallSuccessful(WorkflowExecutionActor.scala:314)
    at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor$$anonfun$3.applyOrElse(WorkflowExecutionActor.scala:97)
    at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor$$anonfun$3.applyOrElse(WorkflowExecutionActor.scala:82)
    at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
    at akka.actor.FSM$class.processEvent(FSM.scala:663)
    at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.akka$actor$LoggingFSM$$super$processEvent(WorkflowExecutionActor.scala:33)
    at akka.actor.LoggingFSM$class.processEvent(FSM.scala:799)
    at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.processEvent(WorkflowExecutionActor.scala:33)
    at akka.actor.FSM$class.akka$actor$FSM$$processMsg(FSM.scala:657)
    at akka.actor.FSM$$anonfun$receive$1.applyOrElse(FSM.scala:651)
    at akka.actor.Actor$class.aroundReceive(Actor.scala:484)
    at cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.aroundReceive(WorkflowExecutionActor.scala:33)
    at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
    at akka.actor.ActorCell.invoke(ActorCell.scala:495)
    at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
    at akka.dispatch.Mailbox.run(Mailbox.scala:224)
    at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)```

**Comment by mcovarr** [comment]
I suggested in standup today that a batching solution like that in [WriteMetadataActor](https://github.com/broadinstitute/cromwell/blob/develop/services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala) might help here.  i.e. don't sweep the execution store every single time any job changes status, but instead sweep based on a timer and a `var shouldCheckForRunnableJobs: Boolean` variable.  Whenever a job changes status `shouldCheckForRunnableJobs` would be assigned `true`.  When the timer goes off and if that variable is `true`, check the store for runnable jobs.

**Comment by Horneth** [comment]
Is there an easy way for me to reproduce this ? Sounds like @geoffjentry you're saying that a 200k wide scatter should make it happen ?

**Comment by geoffjentry** [comment]
@horneth Talk to miguel as he was the last to look at this and would have the most up to date information. IIRC where this was left off is that there were a handful of known hotspots, all of the low hanging fruit had been picked and at this point we're talking fundamental changes to how things are being stored/tracked.

As you can see from the numbers used it'll vary by the oomph your computer has but you'll want to find a number closer to the "it starts happening here" point vs some arbitrarily huge number as IIRC some of the initial chokepoints were less bad than deeper ones but at huge numbers were still horrible to wait through.

You'll definitely want to become friends with JProfiler if you're not already friends with it. The lightbend monitor could also be helpful here (although I'm less sure of that for these purposes) but it's probably worth waiting until I'm back for that.

**Comment by Horneth** [comment]
@yfarjoun @meganshand If I point you to a cromwell branch that might fix this problem, is this something you could easily test ?
I did some testing on my own and it's definitely better but hard to tell if it would really solve this without actually testing it for real.
If the answer is "you can run this WDL yourself and check" that's also fine. Just need to have access to all the inputs (and figure out how to get permission to run something on SGE..)

**Comment by meganshand** [comment]
@Horneth I think I can dig up the WDL and try it out. Do I need to change anything in the conf file for the latest version? (I've only used 24 so far)

**Comment by Horneth** [comment]
Thanks ! Based on the Cromwell changelog for 25 I think you should be good to go with the same config. This is the branch: `cromwell-1938`

**Comment by meganshand** [comment]
@Horneth I tried it with call-caching based on file path only, and while it was significantly faster than it used to be, it still got bogged down. I restarted it with call-caching turned off completely, and then things seemed to work quite well. 

It's running over 1000 jobs with no "Cromwell Overhead" in the timing diagram. So I would say this works when call-caching is turned off, but with call-caching it is still slow.

**Comment by geoffjentry** [comment]
My $0.02 here is that this ticket doesn't imply "with call caching". The slowdowns will be completely unrelated and the problem which @yfarjoun doesn't involve call caching. If the test WDL went from "super slow" to "AOK" w/o call caching enabled I'd like to consider this closed. @meganshand WDYT?

**Comment by meganshand** [comment]
I agree.

**Comment by geoffjentry** [comment]
@meganshand Sweet, I'm going to close. Out of curiosity do you have any before/after in terms of how long running the wdl took (again, sans call caching)? I'm trying to get a sense of the real world magnitude of the performance improvement for this.

**Comment by meganshand** [comment]
I don't know about the actual difference. I also have to admit that there is another possible variable here which is that when we ran this originally the farm was free and clear so the jobs ran very quickly. Now the farm is quite busy which is slowing down the workflow, but this is not the fault of cromwell. 

Maybe @yfarjoun would have a better sense of timing increase.

**Comment by geoffjentry** [comment]
@meganshand Fair enough :) As long as it is somewhere in the range of "from intolerable to AOK", which is what my impression has been here, that's enough for me.

**Comment by Horneth** [comment]
@meganshand when call caching was on, you said it was using file paths, what about [call cache copying](https://github.com/broadinstitute/cromwell/blob/develop/core/src/main/resources/reference.conf#L214) ? Is it full copying or (soft/hard) linking ?

**Comment by Horneth** [comment]
Also do you happen to have a thread dump of cromwell being slow with CC on ?

**Comment by meganshand** [comment]
@Horneth Call cache copying seems to be working! It's relatively slow during "Backend is copying Cached Outputs", but the other tasks that run don't get stuck in "Queued in Cromwell" or "Cromwell Overhead". So I'd say that it works with traditional (not just the file path) call caching. 

I don't have a thread dump with file path CC on but I could rerun it.

**Comment by geoffjentry** [comment]
Forgot to close this. @meganshand & @Horneth feel free to continue to chat here :)

**Comment by geoffjentry** [comment]
err i forgot to wait for @Horneth 's PR to e merged ;)

**Comment by meganshand** [comment]
With some help from @Horneth it became clear that I was using the wrong jar when I ran with file path call caching. Everything seems to be working much faster! Thank you!

**Comment by Horneth** [comment]
Cool ! I merged the PR, this fix is now in develop, so I'll close this !
